File: agnt-plugin/cache.js
// PulseMesh JS — Cache Management
// Port of Python cache.js

import fs from 'fs';
import path from 'path';
import { TelemetrySeries } from './models.js';

function safeId(id) {
  return id.replace(/[^a-zA-Z0-9\-_]/g, '_');
}

export function cachePath(cacheDir, profileId) {
  return path.join(cacheDir, `${safeId(profileId)}.json`);
}

export function saveSeries(cacheDir, series) {
  fs.mkdirSync(cacheDir, { recursive: true });
  fs.writeFileSync(cachePath(cacheDir, series.profileId), JSON.stringify(series, null, 2));
}

export function loadSeries(cacheDir, profileId, reason) {
  const p = cachePath(cacheDir, profileId);
  if (!fs.existsSync(p)) return null;
  try {
    const obj = JSON.parse(fs.readFileSync(p, 'utf-8'));
    return new TelemetrySeries({
      profileId: String(obj.profileId),
      provider: String(obj.provider),
      label: String(obj.label),
      sensorName: String(obj.sensorName || 'Cached telemetry'),
      times: obj.times.map(String),
      values: obj.values.map(Number),
      unit: String(obj.unit || ''),
      usedLiveData: false,
      fallbackReason: `cache fallback after ${reason}`,
      sourceUrl: obj.sourceUrl || null,
      metadata: { ...(obj.metadata || {}), cacheSource: p },
    });
  } catch {
    return null;
  }
}

File: src/pulsemesh/cache.py
from __future__ import annotations

from dataclasses import asdict
from pathlib import Path

from .models import TelemetrySeries
from .util import load_json, write_json


def cache_path(cache_dir: Path, profile_id: str) -> Path:
    safe = "".join(ch if ch.isalnum() or ch in {"-", "_"} else "_" for ch in profile_id)
    return cache_dir / f"{safe}.json"


def save_series(cache_dir: Path, series: TelemetrySeries) -> None:
    write_json(cache_path(cache_dir, series.profile_id), asdict(series))


def load_series(cache_dir: Path, profile_id: str, reason: str) -> TelemetrySeries | None:
    path = cache_path(cache_dir, profile_id)
    if not path.exists():
        return None
    try:
        obj = load_json(path)
        return TelemetrySeries(
            profile_id=str(obj["profile_id"]),
            provider=str(obj["provider"]),
            label=str(obj["label"]),
            sensor_name=str(obj.get("sensor_name", "Cached telemetry")),
            times=[str(x) for x in obj.get("times", [])],
            values=[float(x) for x in obj.get("values", [])],
            unit=str(obj.get("unit", "")),
            used_live_data=False,
            fallback_reason=f"cache fallback after {reason}",
            source_url=obj.get("source_url"),
            metadata={**dict(obj.get("metadata", {})), "cache_source": str(path)},
        )
    except Exception:
        return None

